torch/csrc/jit/passes/quantization.cpp

```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+  std::vector<Value*> values_to_observe;
+  std::vector<Value*> input_values;
+
+  // For traversing all blocks in the graph including subblocks.
+  std::stack<Block*> blocks_to_visit;
+
+  // Mark observer nodes for inputs so we dont add observers
+  // for observers while traversing graph
+  std::unordered_set<Node*> observer_for_input;
+
+  // Add observer for external input nodes excluding parameters
+  // These are treated as activation as they vary across batches
+  // and need to be observed.
+
+  // prim::Param nodes do not belong to the graph. Hence the Insert
+  // point is the beginning of graph node. This also safe guards against
+  // observing a potentially mutated value due to some in-place operation
+  for (size_t idx = 0; idx < num_activation_inputs; ++idx) {
+    auto& v = graph->inputs()[idx];
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      input_values.push_back(v);
+    }
+  }
+
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
+      // Skip nodes that we don't need to observe, e.g. 'prim::Constant' or
+      // observer nodes
+      if (!outputsNeedToBeObserved(n) || observer_for_input.count(n) != 0) {
+        continue;
+      }
+
+      // Record all outputs in the values_to_observe - we'll later add observers
+      // for all values from it.
+      for (Value* v : n->outputs()) {
+        values_to_observe.push_back(v);
+      }
+
+      // Schedule subblocks (if any) for visiting.
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
+    }
+  }
+
+  std::vector<std::string> observer_modules_to_remove;
+  std::vector<Node*> nodes_to_destroy;
+
+  for (Value* v : values_to_observe) {
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      if (v->node()->kind() == prim::GetAttr && v->node()->s(c10::attr::name) == "bias") {
+        quantizeBias(module, v);
+      } else {
+        c10::optional<script::Module> observer_module;
+        std::tie(std::ignore, observer_module) = findObserverModule(module, v);
+        if (observer_module) {
+          quantizeTensor(module, v, observer_modules_to_remove, nodes_to_destroy);
+        }
+      }
+    }
+  }
+
+  for (Value* v : input_values) {
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      c10::optional<script::Module> observer_module;
+      std::tie(std::ignore, observer_module) = findObserverModule(module, v);
+      if (observer_module) {
+        quantizeTensor(module, v, observer_modules_to_remove, nodes_to_destroy, false);
+      }
+    }
+  }
+
+  // Destroy observer forward calls
+  for (auto& n: nodes_to_destroy) {
+    n->destroy();
+  }
+
+  // NOTE: Remove observer module does not work right now, we'll return
+  // the module with observer modules as a temporary workaround
+  // TODO: remove observer modules after we have a remove_module API
+
+  return module;
```

Comment:
I don't know if it makes sense to return module as well when we're mutating it in-place.